#' countFinalRegions
#' @description count reads falling within the final regions.
#'
#' @param regionsGRanges a GRanges objects representing the peaks to compute
#' the coverage, with a "k-carriers" mcols.
#' (tipically generated by finalRegions function).
#' @param readsFilePath the filepath of bam or bed files necessary to compute
#' the coverage.
#' @param fileType the file type of the input files.
#' @param minCarriers minimum number of carriers (samples).
#' @param genomeName code name of the genome of reads files (i.e. "mm9").
#' @param onlyStdChrs a flag indicating if to keep only the standard chromosomes
#' @param ignStrandSO a flag indicating if to ignore the reads strand.
#' (see GenomicAlignments::summarizeOverlaps).
#' @param modeSO the mode to use, default is "Union".
#' (see GenomicAlignments::summarizeOverlaps).
#' @param saveFlag a flag indicating if to save the results.
#' @param savePath the path where to store the results.
#' @param verbose verbose output.
#' @return A Matrix containing read counts with regions as rows and samples as
#' columns.
#' @export
#' @importFrom GenomicAlignments summarizeOverlaps
#' @importFrom SummarizedExperiment assay
# @examples TBW
countFinalRegions <- function(regionsGRanges, readsFilePath=NULL,
                                fileType=c("bam", "bed"),
                                minCarriers=2,
                                genomeName=NULL,
                                onlyStdChrs=FALSE,
                                ignStrandSO=TRUE,
                                modeSO="Union",
                                saveFlag=FALSE,
                                savePath="finalRegions",
                                verbose=FALSE)
{
    match.arg(fileType)
    stopifnot(is(regionsGRanges, "GRanges"))
    if(is.null(readsFilePath)) {stop("readsFilePath cannot be NULL!")}

    idxK <- which(regionsGRanges$`k-carriers` >= minCarriers)
    regionsGRanges <- regionsGRanges[idxK,]

    readsFiles <- list.files(readsFilePath, full.names=TRUE, pattern=fileType)
    idxBai <- grep("bai", readsFiles)
    if(length(idxBai) > 0) readsFiles <- readsFiles[-idxBai]
    if(verbose) message("Final regions on ", length(readsFiles), "\n")
    summRegMat <- sapply(readsFiles, function(file)
    {
        fileReads <- constructBedRanges(filename=as.character(file),
                                        filetype=fileType,
                                        genomeName=genomeName,
                                        onlyStdChrs=onlyStdChrs)
        regionsGRanges <- setGRGenomeInfo(GRanges=regionsGRanges,
                                            genomeName=genomeName)
        summReg <- GenomicAlignments::summarizeOverlaps(
                                                    features=regionsGRanges,
                                                    reads=fileReads,
                                                    ignore.strand=ignStrandSO,
                                                    mode=modeSO)
        return(SummarizedExperiment::assay(summReg))
    })

    regionsRN <- paste0(regionsGRanges@seqnames, ":",
                        start(regionsGRanges), "-",
                        end(regionsGRanges))

    rownames(summRegMat) <- regionsRN
    if(saveFlag)
    {
        ## TBW
    }

    return(summRegMat)
}
