#' countFinalRegions
#' @description count reads falling within the final regions.
#'
#' @param regionsGRanges a GRanges objects representing the peaks to compute
#'                       the coverage, with a "k-carriers" mcols.
#'                       (tipically generated by finalRegions function)
#' @param readsFilePath the filepath of bam or bed files necessary to compute
#'                      the coverage
#' @param fileType the fileType of the input files
#' @param minCarriers minimum number of carriers (samples)
#' @param genomeName code name of the genome of reads files (optional)
#' @param onlyStdChrs a flag indicating if to keep only the standard chromosomes
#' @param saveFlag a flag indicating if to save the results
#' @param savePath the path where to store the results
#'
#' @return Matrix containing read counts with regions as rows and samples as
#'     columns.
#' @export
#' @importFrom GenomicAlignments summarizeOverlaps
#' @importFrom SummarizedExperiment assay
# @examples TBW
countFinalRegions <- function(regionsGRanges, readsFilePath=NULL,
                              fileType=c("bam", "bed"),
                              minCarriers=2,
                              #minCoverage=0,
                              genomeName=NULL,
                              onlyStdChrs=FALSE,
                              saveFlag=FALSE,
                              savePath="finalRegions")
{
    match.arg(fileType)
    stopifnot(is(regionsGRanges, "GRanges"))
    if(is.null(readsFilePath)) {stop("readsFilePath cannot be NULL!")}

    idxK <- which(regionsGRanges$`k-carriers` >= minCarriers)
    regionsGRanges <- regionsGRanges[idxK,]

    readsFiles <- list.files(readsFilePath, full.names=TRUE, pattern=fileType)
    idxBai <- grep("bai", readsFiles)
    if(length(idxBai) > 0) readsFiles <- readsFiles[-idxBai]

    summRegMat <- sapply(readsFiles, function(file)
    {
        fileReads <- constructBedRanges(filename=as.character(file),
                                        filetype=fileType,
                                        genomeName=genomeName,
                                        onlyStdChrs=onlyStdChrs)
        summReg <- GenomicAlignments::summarizeOverlaps(features=regionsGRanges,
                                                        reads=fileReads)
        return(SummarizedExperiment::assay(summReg))
    })

    rownames(summRegMat) <- names(regionsGRanges)


    if(saveFlag)
    {

    }

    return(summRegMat)
}
