#' countFinalRegions
#' @description count reads falling within the final regions.
#'
#' @param regionsGRanges a GRanges objects representing the peaks to compute
#' the coverage, with a "k-carriers" mcols.
#' (tipically generated by finalRegions function).
#' @param readsFilePath the filepath of bam or bed files necessary to compute
#' the coverage.
#' @param fileType the file type of the input files.
#' @param minCarriers minimum number of carriers (samples).
#' @param genomeName code name of the genome of reads files (i.e. "mm9").
#' @param onlyStdChrs a flag indicating if to keep only the standard chromosomes
#' @param ignStrandSO a flag indicating if to ignore the reads strand.
#' (see GenomicAlignments::summarizeOverlaps).
#' @param modeSO the mode to use, default is "Union".
#' (see GenomicAlignments::summarizeOverlaps).
#' @param saveFlag a flag indicating if to save the results.
#' @param savePath the path where to store the results.
#' @param verbose verbose output.
#'
#' @return A Matrix containing read counts with regions as rows and samples as
#' columns.
#' @export
#' @importFrom GenomicAlignments summarizeOverlaps
#' @importFrom SummarizedExperiment assay SummarizedExperiment
#' @importFrom BiocGenerics start end
#' @examples
#' filename <- system.file("extdata/regions/regions_zt20_minK4GR.RDS",
#'                         package="DEScan2")
#' regionsGR <- readRDS(file=)
#' finalRegions <- countFinalRegions(regionsGRanges=regionsGR,
#'                                 readsFilePath="inst/extdata/bam",
#'                                 fileType="bam",
#'                                 minCarriers=1,
#'                                 genomeName="mm9",
#'                                 onlyStdChrs=TRUE,
#'                                 ignStrandSO=TRUE, saveFlag=TRUE,
#'                                 savePath="testData/regions/prova/",
#'                                 verbose=TRUE)
#' head(countFinalRegions)
countFinalRegions <- function(regionsGRanges, readsFilePath=NULL,
                                fileType=c("bam", "bed"),
                                minCarriers=2,
                                genomeName=NULL,
                                onlyStdChrs=FALSE,
                                ignStrandSO=TRUE,
                                modeSO="Union",
                                saveFlag=FALSE,
                                savePath="finalRegions",
                                verbose=TRUE)
{
    match.arg(fileType)
    stopifnot(is(regionsGRanges, "GRanges"))
    if(is.null(readsFilePath)) {stop("readsFilePath cannot be NULL!")}

    idxK <- which(regionsGRanges$`k-carriers` >= minCarriers)
    regionsGRanges <- regionsGRanges[idxK,]

    regionsGRanges <- setGRGenomeInfo(GRanges=regionsGRanges,
                                      genomeName=genomeName)

    if(fileType == "bam")
    {
        readsFiles <- list.files(path=readsFilePath, full.names=TRUE,
                                pattern=".bam$")
    }
    else
    {
        readsFiles <- list.files(path=readsFilePath, full.names=TRUE,
                                 pattern=".bed$")
    }
    if(length(readsFiles) == 0 ) stop("No reads files found!")
    if(verbose) message("Final regions on ", length(readsFiles), " files.")

    fileReadsList <- lapply(readsFiles, function(file)
    {
        fileReads <- constructBedRanges(filename=as.character(file),
                                        filetype=fileType,
                                        genomeName=genomeName,
                                        onlyStdChrs=onlyStdChrs)
        return(fileReads)
    })
    names(fileReadsList) <- readsFiles

    summRegMat <- sapply(fileReadsList, function(fileReads)
    {
        summReg <- GenomicAlignments::summarizeOverlaps(
            features=regionsGRanges,
            reads=fileReads,
            ignore.strand=ignStrandSO,
            mode=modeSO)
        return(SummarizedExperiment::assay(summReg))
    })

    regionsRN <- paste0(regionsGRanges@seqnames, ":",
                        BiocGenerics::start(regionsGRanges), "-",
                        BiocGenerics::end(regionsGRanges))

    rownames(summRegMat) <- regionsRN
    if(saveFlag)
    {
        datename <- paste0(strsplit(gsub(pattern=":", replacement=" ",
                                         date()), " ")[[1]], collapse="_")
        filename <- paste0("regions_", datename, "_minK", minCarriers,
                           "_mso", modeSO, ".tsv")
        write.table(x=summRegMat, file=paste0(savePath, filename), quote=FALSE,
                    sep="\t")
        if(verbose) message("file ", filename, " saved on disk!")
    }
    rownames(summRegMat) <- names(regionsGRanges)
    summExp <- SummarizedExperiment::SummarizedExperiment(assays=summRegMat,
                                                rowRanges=regionsGRanges,
                                                colData=data.frame(readsFiles))
    return(summExp)
}
